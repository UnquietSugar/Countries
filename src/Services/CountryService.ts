export default class CountryService {

  private path: string = 'https://restcountries.com/v3.1/';

  public async getAllCountries() {
    const res = await fetch(this.path + 'all');
    return res.json();
  }

  public async getByLanguage(lang: string) {
    const res = await fetch(this.path + `lang/${lang}`);
    const data = await res.json()
    // const updated = { ...data }
    const commonLang: string = data[0].languages[lang];
    return { data: data, lang: commonLang };
  }
}