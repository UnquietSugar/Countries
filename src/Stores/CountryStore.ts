import { observable, action } from 'mobx';
import { ICountriesByLanguage } from '../Interfaces/ICountriesByLanguage';
import { ICountry } from '../Interfaces/ICountry';
import CountryService from '../Services/CountryService';



export default class CountryStore {

  private countryService: CountryService;
  @observable public countries: ICountry[] = [];
  @observable public currentCountries: ICountry[] = [];
  @observable public langArray: string[] = [];
  @observable public currentLangArr: string[] = [];
  @observable public currentCountriesByLang: ICountriesByLanguage[] = [];

  constructor(_countryService: CountryService) {
    this.countryService = _countryService;
    this.onInitialize();
  };

  private async onInitialize() {
    await this.getAllCountries();
    this.extractLanguageKeysToArray();
  }

  @action private async getAllCountries() {
    this.countries = await this.countryService.getAllCountries();
    this.assignOriginalCountriesArray();
  };

  @action public assignOriginalCountriesArray() {
    this.currentCountries = this.countries;
  };

  @action public sortByName() {
    this.currentCountries = this.countries.sort((a: ICountry, b: ICountry) => a.name.common.localeCompare(b.name.common));
  };

  @action public sortByPopulation() {
    this.currentCountries = this.countries.sort((a: ICountry, b: ICountry) => a.population - b.population);
  };

  @action public sortByArea() {
    this.currentCountries = this.countries.sort((a: ICountry, b: ICountry) => a.area - b.area);
  };

  @action public getCountriesByLanguage(page: number, itemsPerPage: number) {
    this.currentLangArr = this.langArray.slice(page * itemsPerPage, (itemsPerPage * page) + itemsPerPage);

    this.currentCountriesByLang = [];

    this.currentLangArr.forEach(async (lang: string) => {
      this.currentCountriesByLang.push(await this.countryService.getByLanguage(lang));
    });
  };

  public findCountryWithSmallestArea() {
    return this.countries.reduce((a: ICountry, b: ICountry) => a.area < b.area ? a : b);
  };

  public findCountryWithBiggestArea() {
    return this.countries.reduce((a: ICountry, b: ICountry) => a.area > b.area ? a : b);
  };

  public async extractLanguageKeysToArray() {
    let arr: string[][] = [];

    this.countries.forEach((c: ICountry) => {
      if (!!c.languages) {
        arr.push(Object.keys(c.languages))
      };
    });

    this.langArray = [...new Set(arr.flat())];
  };

};