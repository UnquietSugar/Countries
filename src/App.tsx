import React, { FC } from 'react';
import DataTable from './Components/DataTable';
import Loader from './Components/Loader';
import Navbar from './Components/Navbar';
import Summary from './Components/Summary';
import LangTable from './Components/LangTable';
import { CountryStore } from './Stores/CountryStore';
import { inject, observer } from 'mobx-react';
import { Box } from '@mui/system';
import { BrowserRouter as Router, Route, Routes } from "react-router-dom";

interface IAppProps {
  countryStore?: CountryStore;
}

const App: FC<IAppProps> = inject('countryStore')(observer(({ countryStore }) => {

  const countTable = () => {
    return (
      <Box padding={10}>
        {countryStore?.countries.length > 0 ?
          <>
            <DataTable />
            <Summary />
          </>
          :
          <Loader />}
      </Box>);
  }

  const langTable = () => {
    return (
      <Box padding={10}>
        {countryStore?.countries.length > 0 ?
          <LangTable />
          :
          <Loader />}
      </Box>
    )
  }


  return (

    <Router>
      <Navbar />
      <Routes>
        <Route path="/" element={countTable()} />
        <Route path="/lang-table" element={langTable()} />
      </Routes>
    </Router>

  );

}));


export default App;
