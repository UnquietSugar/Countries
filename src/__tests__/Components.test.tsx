import { Provider } from 'mobx-react';
import renderer from 'react-test-renderer';
import App from '../App';
import { CountryService } from '../Services/CountryService';
import { CountryStore } from '../Stores/CountryStore';


const countryService = new CountryService();
const countryStore = new CountryStore(countryService);

describe('App.tsx tests', () => {

  it('should render App.tsx, and have two elements inside', () => {
    const tree: any = renderer.create(<Provider countryStore={countryStore}><App /></Provider>).toJSON();
    expect(tree).toMatchSnapshot();
    expect(tree.length).toEqual(2);
  });


});