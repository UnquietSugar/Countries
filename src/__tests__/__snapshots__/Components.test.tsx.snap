// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App.tsx tests should render App.tsx, and have two elements inside 1`] = `
Array [
  <header
    className="MuiPaper-root MuiPaper-elevation MuiPaper-elevation4 MuiAppBar-root MuiAppBar-colorPrimary MuiAppBar-positionStatic makeStyles-navBar-1 css-hip9hq-MuiPaper-root-MuiAppBar-root"
  >
    <div
      className="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular css-1t29gy6-MuiToolbar-root"
    >
      <h4
        className="MuiTypography-root MuiTypography-h4 makeStyles-title-2 css-5lbw0b-MuiTypography-root"
      >
        Countries
      </h4>
      <a
        href="/"
        onClick={[Function]}
        style={
          Object {
            "marginRight": 50,
            "textDecoration": "none",
          }
        }
      >
        <button
          aria-selected={true}
          className="MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButtonBase-root  css-11gc9ze-MuiButtonBase-root-MuiButton-root"
          disabled={false}
          onBlur={[Function]}
          onContextMenu={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={0}
          type="button"
        >
          Countries Table
        </button>
      </a>
      <a
        href="/lang-table"
        onClick={[Function]}
        style={
          Object {
            "marginRight": 50,
            "textDecoration": "none",
          }
        }
      >
        <button
          aria-selected={true}
          className="MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButtonBase-root  css-11gc9ze-MuiButtonBase-root-MuiButton-root"
          disabled={false}
          onBlur={[Function]}
          onContextMenu={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={0}
          type="button"
        >
          Languages Table
        </button>
      </a>
      <div
        className="MuiBox-root css-ece9u5"
      >
        <div
          className="MuiFormControl-root MuiFormControl-fullWidth css-i6wma6-MuiFormControl-root"
        >
          <label
            className="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-filled MuiFormLabel-root MuiFormLabel-colorPrimary css-e4w4as-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink={false}
            id="demo-simple-select-label"
          >
            Sort by...
          </label>
          <div
            className="MuiFilledInput-root MuiFilledInput-underline MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl css-cio0x1-MuiInputBase-root-MuiFilledInput-root"
            onClick={[Function]}
          >
            <div
              aria-expanded="false"
              aria-haspopup="listbox"
              aria-labelledby="demo-simple-select-label demo-simple-select"
              className="MuiSelect-select MuiSelect-filled MuiFilledInput-input MuiInputBase-input css-d9oaum-MuiSelect-select-MuiInputBase-input-MuiFilledInput-input"
              id="demo-simple-select"
              onBlur={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onMouseDown={[Function]}
              role="button"
              tabIndex={0}
            >
              <span
                className="notranslate"
                dangerouslySetInnerHTML={
                  Object {
                    "__html": "&#8203;",
                  }
                }
              />
            </div>
            <input
              aria-hidden={true}
              className="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
              disabled={false}
              onAnimationStart={[Function]}
              onChange={[Function]}
              required={false}
              tabIndex={-1}
              value=""
            />
            <svg
              aria-hidden={true}
              className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconFilled css-hfutr2-MuiSvgIcon-root-MuiSelect-icon"
              data-testid="ArrowDropDownIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M7 10l5 5 5-5z"
              />
            </svg>
          </div>
        </div>
      </div>
    </div>
  </header>,
  <div
    className="MuiBox-root css-tvxkav"
  >
    <div
      className="MuiBox-root css-1d60z6w"
    >
      <span
        className="MuiCircularProgress-root MuiCircularProgress-indeterminate MuiCircularProgress-colorPrimary css-18lrjg1-MuiCircularProgress-root"
        role="progressbar"
        style={
          Object {
            "height": 40,
            "width": 40,
          }
        }
      >
        <svg
          className="MuiCircularProgress-svg css-1idz92c-MuiCircularProgress-svg"
          viewBox="22 22 44 44"
        >
          <circle
            className="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate css-176wh8e-MuiCircularProgress-circle"
            cx={44}
            cy={44}
            fill="none"
            r={20.2}
            strokeWidth={3.6}
            style={Object {}}
          />
        </svg>
      </span>
    </div>
  </div>,
]
`;
